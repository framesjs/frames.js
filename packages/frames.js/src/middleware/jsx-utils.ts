import { Children as ReactChildren, Fragment, createElement } from "react";

export type SerializedNode =
  | {
      type: string;
      // eslint-disable-next-line @typescript-eslint/no-explicit-any -- we need to handle any type
      props: Record<string, any> & {
        children: SerializedNode[];
      };
    }
  | (string | number | boolean | null | undefined);

export function deserializeJsx(
  serialized: SerializedNode[]
): React.ReactElement {
  // Map over each serialized node and convert it to a React element or a text node
  const elements = serialized.map((node) => {
    if (typeof node === "object" && node !== null && "type" in node) {
      // It's an element object with type and props
      const children = node.props.children
        ? deserializeJsx(node.props.children)
        : [];
      const { children: _, ...restProps } = node.props;
      return createElement(
        node.type === "Symbol(react.fragment)" ? Fragment : node.type,
        {
          ...restProps,
          key: "key" in node ? (node.key as string) : Math.random().toString(),
        },
        children
      );
    } else if (node !== null && node !== undefined) {
      // It's a primitive value, so just return it as is (React can render strings and numbers directly)
      return node;
    }
    // Null or undefined children are valid in React and can simply be ignored
    return null;
  });

  // If there's only one top-level node, return it directly, otherwise return the array
  return (elements.length === 1 ? elements[0] : elements) as JSX.Element;
}
/* eslint-disable @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment -- we need to handle any type */
function serializeChild(
  child: React.ReactElement | null | undefined | boolean | string
): SerializedNode {
  if (typeof child !== "object" || !child) {
    return child;
  }

  if ("props" in child) {
    let serialized: React.ReactElement = { ...child };

    if ("children" in child.props && child.props.children) {
      const childChildren = serializeJsx(
        child.props.children as React.ReactElement
      );

      serialized = {
        ...serialized,
        props: {
          ...serialized.props,
          children: childChildren,
        },
      };
    }

    // If the child is a functional component, evaluate it and return the evaluation result
    if (typeof child.type === "function") {
      const evaluated = (child.type as CallableFunction)(
        child.props
      ) as React.ReactElement | null;

      if (evaluated === null || typeof evaluated !== "object") {
        return evaluated;
      }

      const serializedEvaluated = serializeChild({ ...evaluated });

      if (
        serializedEvaluated === null ||
        typeof serializedEvaluated !== "object"
      ) {
        return serializedEvaluated;
      }

      serialized = { ...serializedEvaluated, key: Math.random().toString() };
    }

    return {
      type: serialized.type.toString(),
      props: serialized.props,
    };
  }
}

export function serializeJsx(children: React.ReactElement): SerializedNode[] {
  return ReactChildren.map(children, serializeChild);
}
