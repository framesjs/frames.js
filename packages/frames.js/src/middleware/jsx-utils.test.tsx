import { serializeJsx, deserializeJsx } from "./jsx-utils";

describe("serializeJsx", () => {
  it("correctly serializes an HTML element", () => {
    const element = <div />;
    expect(serializeJsx(element)).toEqual([
      {
        type: "div",
        props: {},
      },
    ]);
  });

  it("correctly serializes functional component", () => {
    function Test(): string {
      return "Hello";
    }

    expect(serializeJsx(<Test />)).toEqual(["Hello"]);
  });

  it("correctly serializes a Fragment", () => {
    expect(serializeJsx(<>Test</>)).toEqual([
      {
        props: {
          children: ["Test"],
        },
        type: "Symbol(react.fragment)",
      },
    ]);
  });

  it("correctly serializes a JSX element with nested array of children that are functional components", () => {
    function Test(): React.ReactElement {
      return <div>Test</div>;
    }

    const image = (
      <div>
        <Test />
        <Test />
      </div>
    );

    expect(serializeJsx(image)).toEqual([
      {
        type: "div",
        props: {
          children: [
            {
              type: "div",
              props: {
                children: ["Test"],
              },
            },
            {
              type: "div",
              props: {
                children: ["Test"],
              },
            },
          ],
        },
      },
    ]);
  });

  it("correctly serializes nested tree", () => {
    const element = (
      <div>
        <span>Test</span>
        <span>Test 2</span>
        <div>
          <div>
            <span>Test</span>
            <span>Test 2</span>
          </div>
        </div>
      </div>
    );

    expect(serializeJsx(element)).toEqual([
      {
        type: "div",
        props: {
          children: [
            {
              type: "span",
              props: {
                children: ["Test"],
              },
            },
            {
              type: "span",
              props: {
                children: ["Test 2"],
              },
            },
            {
              type: "div",
              props: {
                children: [
                  {
                    type: "div",
                    props: {
                      children: [
                        { type: "span", props: { children: ["Test"] } },
                        { type: "span", props: { children: ["Test 2"] } },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    ]);
  });

  it("correctly serializes nested tree with fragments", () => {
    const element = (
      <div>
        <span>Test</span>
        <>
          <span>Test2</span>
          <span>Test3</span>
        </>
      </div>
    );

    expect(serializeJsx(element)).toEqual([
      {
        type: "div",
        props: {
          children: [
            {
              type: "span",
              props: {
                children: ["Test"],
              },
            },
            {
              type: "Symbol(react.fragment)",
              props: {
                children: [
                  {
                    type: "span",
                    props: {
                      children: ["Test2"],
                    },
                  },
                  {
                    type: "span",
                    props: {
                      children: ["Test3"],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    ]);
  });
});

describe("deserializeJsx", () => {
  it("correctly deserializes an HTML element", () => {
    const element = <div />;
    expect(deserializeJsx(serializeJsx(element))).toMatchSnapshot();
  });

  it("correctly deserializes functional component", () => {
    function Test(): string {
      return "Hello";
    }

    expect(deserializeJsx(serializeJsx(<Test />))).toMatchSnapshot();
  });

  it("correctly deserializes a Fragment", () => {
    expect(deserializeJsx(serializeJsx(<>Test</>))).toMatchSnapshot();
  });

  it("correctly deserializes nested tree", () => {
    const element = (
      <div>
        <span>Test</span>
        <span>Test2</span>
      </div>
    );

    expect(deserializeJsx(serializeJsx(element))).toMatchSnapshot();
  });

  it("is backwards compatible with old serialized format", () => {
    expect(
      deserializeJsx([
        {
          props: {
            children: [
              { props: { children: ["Test"] }, type: "div" },
              { props: { children: ["Test"] }, type: "div" },
            ],
          },
          type: "div",
        },
      ])
    ).toMatchSnapshot();
    expect(
      deserializeJsx([
        {
          props: {
            children: ["Test"],
            style: { color: "red" },
            tw: "flex bg-gray-500",
          },
          type: "div",
        },
      ])
    ).toMatchSnapshot();
  });
});
