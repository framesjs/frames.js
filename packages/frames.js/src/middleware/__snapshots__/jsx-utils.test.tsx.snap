// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deserializeJsx correctly deserializes a Fragment 1`] = `
<React.Fragment>
  Test
</React.Fragment>
`;

exports[`deserializeJsx correctly deserializes an HTML element 1`] = `<div />`;

exports[`deserializeJsx correctly deserializes functional component 1`] = `"Hello"`;

exports[`deserializeJsx correctly deserializes nested tree 1`] = `
<div>
  <span>
    Test
  </span>
  <span>
    Test2
  </span>
</div>
`;

exports[`deserializeJsx is backwards compatible with old serialized format 1`] = `
<div>
  <div>
    Test
  </div>
  <div>
    Test
  </div>
</div>
`;

exports[`deserializeJsx is backwards compatible with old serialized format 2`] = `
<div
  style={
    {
      "color": "red",
    }
  }
  tw="flex bg-gray-500"
>
  Test
</div>
`;
